File: Python/Sequences/(56)MergeIntervals.py
# Sorting
# Time: O(nlogn) -> Sort is O(nlogn)
# Space: O(logn)
class Solution:
    def merge(self, intervals: List[List[int]]) -> List[List[int]]:
        answer = []
        
        intervals.sort()
        
        for item in intervals:
            if not answer or answer[-1][1] < item[0]:
                answer.append(item)
            else:
                answer[-1][1] = max(answer[-1][1], item[1])
            
        return answer
        

File: C++/Sequences/(56)MergeIntervals.cpp
// Sorting
// Time: O(nlogn) -> Sort is O(nlogn)
// Space: O(logn)
class Solution {
public:
    vector<vector<int>> merge(vector<vector<int>>& intervals) {
        vector<vector<int>> answer;
        
        sort(intervals.begin(), intervals.end());
        
        for (int i = 0; i < intervals.size(); i++) {
            if (answer.empty() || answer.back()[1] < intervals[i][0])
                answer.push_back(intervals[i]);
            else
                answer.back()[1] = max(answer.back()[1], intervals[i][1]);
        }
        
        return answer;
    }
};